### PR TITLE
boards: mimxrt1170_evk_cm7: Enable USDHC on RT1170 CM7 Core

### PR DESCRIPTION
**Pull request by danieldegrasse**
Enables USDHC peripheral on RT1170 CM7 core. Since the RT1170 EVK does not have a DETECT gpio for the SD card wired by default, this PR adds support for card detection using DAT3 pin.

### CODE REVIEW

File: boards/arm/mimxrt1170_evk/pinmux.c [hunk 1]
```diff
@@ -34,59 +34,17 @@ static gpio_pin_config_t cam_pwdn_config = {
...
-		IOMUXC_SetPinMux(IOMUXC_GPIO_SD_B1_02_USDHC1_DATA0, 1U);
-		IOMUXC_SetPinMux(IOMUXC_GPIO_SD_B1_03_USDHC1_DATA1, 1U);
-		IOMUXC_SetPinMux(IOMUXC_GPIO_SD_B1_04_USDHC1_DATA2, 1U);
-		IOMUXC_SetPinMux(IOMUXC_GPIO_SD_B1_05_USDHC1_DATA3, 1U);
-		IOMUXC_GPR->GPR43 = ((IOMUXC_GPR->GPR43
-			& (~(IOMUXC_GPR_GPR43_GPIO_MUX3_GPIO_SEL_HIGH_MASK)))
-			| IOMUXC_GPR_GPR43_GPIO_MUX3_GPIO_SEL_HIGH(0x8000U));
+	if (pullup) {
+		/* Set pin config to pull up */
+		IOMUXC_SetPinConfig(IOMUXC_GPIO_SD_B1_05_USDHC1_DATA3, 0x04U);
```
Comment (dleach02) [inline boards/arm/mimxrt1170_evk/pinmux.c:43]: Can we reference the define instead of a hard number here? 

Same in each of the other SetPinConfigs below

File: drivers/disk/usdhc.c [hunk 2]
```diff
@@ -1135,10 +1146,12 @@ static int usdhc_wait_cmd_done(struct usdhc_priv *priv,
...
 		}
 
-		if ((int_status & USDHC_INT_TUNING_ERR_FLAG) ||
-			(int_status & USDHC_INT_CMD_ERR_FLAG)) {
+		if (int_status & USDHC_INT_TUNING_ERR_FLAG) {
 			error = -EIO;
 		}
+		if (int_status & USDHC_INT_CMD_ERR_FLAG) {
+			error = -ETIMEDOUT;
+		}
```
Comment (dleach02) [inline drivers/disk/usdhc.c:1154]: is EIO and ETIMEDOUT handled differently by the caller? Curious why this change?

File: drivers/disk/usdhc.c [hunk 3]
```diff
@@ -2596,6 +2611,34 @@ static int usdhc_sd_init(struct usdhc_priv *priv)
...
+		/* Call board specific function to pull down DAT3 */
+		imxrt_usdhc_dat3_pull(pullup);
+		/* Power off the card to clear DAT3 legacy status */
+		if (priv->pwr_gpio) {
+			ret = gpio_pin_set(priv->pwr_gpio, priv->config->pwr_pin, 0);
+			if (ret) {
+				return ret;
+			}
+			/* Delay for card power off to complete */
+			usdhc_millsec_delay(1000);
```
Comment (dleach02) [inline drivers/disk/usdhc.c:2632]: a bit of a magic number here. Should this be an internal Kconfig value?

File: drivers/disk/usdhc.c [hunk 4]
```diff
@@ -2635,10 +2683,23 @@ static int usdhc_board_access_init(struct usdhc_priv *priv)
...
 	if (!priv->detect_gpio) {
 		LOG_INF("USDHC detection other than GPIO");
-		/* DATA3 does not monitor card insertion */
-		base->PROT_CTRL &= ~USDHC_PROT_CTRL_D3CD_MASK;
+		if (config->detect_dat3) {
+			base->PROT_CTRL |= USDHC_PROT_CTRL_D3CD_MASK;
+			/* Pull down DAT3 */
+			usdhc_dat3_pull(USDHC_DAT3_PULL_DOWN, priv);
+			/* Delay to allow pin state to change */
+			usdhc_millsec_delay(100);
```
Comment (dleach02) [inline drivers/disk/usdhc.c:2691]: magic number... do we have some reference on what validates this?

File: drivers/disk/usdhc.c [hunk 5]
```diff
@@ -2596,6 +2611,34 @@ static int usdhc_sd_init(struct usdhc_priv *priv)
...
+{
+	int ret = 0U;
+
+	if (pullup) {
+		/* Just call the board specific function (registered in SOC file) */
+		imxrt_usdhc_dat3_pull(pullup);
+	} else {
+		/* Call board specific function to pull down DAT3 */
+		imxrt_usdhc_dat3_pull(pullup);
+		/* Power off the card to clear DAT3 legacy status */
```
Comment (dleach02) [inline drivers/disk/usdhc.c:2625]: doing the same function call at beginning of each sides of if statement. Why not pull the function call out and just do a 

```
imxrt_usdhc_dat3_pull(pullup);

if (!pullup) {
...
}
```

File: soc/arm/nxp_imx/rt/soc_rt11xx.c [hunk 6]
```diff
@@ -393,6 +401,52 @@ static ALWAYS_INLINE void clock_init(void)
...
+#if CONFIG_DISK_DRIVER_SDMMC
+
+/* Usdhc driver needs to re-configure pinmux
+ * Pinmux depends on board design.
+ * From the perspective of Usdhc driver,
+ * it can't access board specific function.
+ * So SoC provides this for board to register
+ * its usdhc pinmux and for usdhc to access
+ * pinmux.
+ */
```
Comment (dleach02) [inline soc/arm/nxp_imx/rt/soc_rt11xx.c:413]: Is there an opportunity here to express this via a board level DTS instead of a board level function?

File: drivers/disk/usdhc.c [hunk 7]
```diff
@@ -840,11 +845,17 @@ static int usdhc_Internal_dma_cfg(struct usdhc_priv *priv,
 		base->ADMA_SYS_ADDR = (uint32_t)(dma_cfg->adma_table);
 	}
 
+#if (defined(FSL_FEATURE_USDHC_HAS_NO_RW_BURST_LEN) && FSL_FEATURE_USDHC_HAS_NO_RW_BURST_LEN)
+	/* select DMA mode and config the burst length */
+	base->PROT_CTRL &= ~(USDHC_PROT_CTRL_DMASEL_MASK);
+	base->PROT_CTRL |= USDHC_PROT_CTRL_DMASEL(dma_cfg->dma_mode);
```
Comment (dleach02) [inline drivers/disk/usdhc.c:851]: Comment suggest setting burst length but is this really setting a burst length since this is inside the "HAS_NO_RW_BURST_LEN"?

File: boards/arm/mimxrt1170_evk/pinmux.c (hunk 1 above)
Comment (danieldegrasse) [reply boards/arm/mimxrt1170_evk/pinmux.c]: Fixed to use the define for all SDMMC pins

File: drivers/disk/usdhc.c (hunk 2 above)
Comment (danieldegrasse) [reply drivers/disk/usdhc.c]: They aren't to my knowledge. This change slipped in from debugging issues with the 1.8V support in the driver (since the logged return code can be useful). I've removed it.

File: drivers/disk/usdhc.c (hunk 3 above)
Comment (danieldegrasse) [reply drivers/disk/usdhc.c]: This is pulled from the board specific SDMMC file. I changed the delay to be a Kconfig value, and the RT1170 EVK defconfig sets it to 1000.

File: drivers/disk/usdhc.c (hunk 4 above)
Comment (danieldegrasse) [reply drivers/disk/usdhc.c]: The SDK has a delay placed there of 1 ms, I'm not sure where I pulled 100ms from. I've removed the delay and DAT3 card detection still works fine.

File: drivers/disk/usdhc.c (hunk 5 above)
Comment (danieldegrasse) [reply drivers/disk/usdhc.c:2621]: Fixed in latest push

File: drivers/disk/usdhc.c (hunk 7 above)
Comment (danieldegrasse) [reply drivers/disk/usdhc.c:851]: It isn't, I missed that change. Fixed in latest push

File: soc/arm/nxp_imx/rt/soc_rt11xx.c (hunk 6 above)
Comment (danieldegrasse) [reply soc/arm/nxp_imx/rt/soc_rt11xx.c:413]: In theory I think this could be moved to a board level DTS, but I believe that would require a pinmux driver, since as it stands we only directly use the IOMUXC calls in board level `pinmux.c` files. We could pull the DAT3 pin specifically into the DTS, and use the IOMUXC functions in the usdhc driver, but that seems like we are reimplementing part of the pinmux functionality

File: drivers/disk/usdhc.c (hunk 4 above)
Comment (dleach02) [reply drivers/disk/usdhc.c]: not necessarily requesting to remove it... just wanted to understand where it was coming from. I have a personal bias against magic numbers. Just find it useful to have comments/references to validate their use for a future reader of the code.

**Review by dleach02 — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by cfriedt** [comment]
@danieldegrasse - please rebase / fix conflicts

**Comment by danieldegrasse** [comment]
@cfriedt Rebase done, thanks for the heads up
